### PR TITLE
HPCC-17900 Ability to capture git status through eclcc

### PR DESCRIPTION
**Pull request by richardkchapman**
New option -checkDirty stores git information into archive (if available).

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
Tested manually

### CODE REVIEW

**Comment by richardkchapman** [comment]
@afishbeck fyi 
@ghalliday Please review/merge

File: ecl/eclcc/eclcc.cpp [hunk 1]
```diff
@@ -1723,7 +1766,7 @@ void EclCC::processReference(EclCompileInstance & instance, const char * queryAt
     const char * outputFilename = instance.outputFilename;
 
     instance.wu.setown(createLocalWorkUnit(NULL));
-    if (optArchive || optGenerateDepend || optSaveQueryArchive)
+    if (optArchive || optGenerateDepend || optSaveQueryArchive || optCheckDirty)
         instance.archive.setown(createAttributeArchive());
```
Comment (ghalliday) [inline ecl/eclcc/eclcc.cpp]: not sure optCheckDirty should create an archive.

File: ecl/hql/hqlcollect.cpp [hunk 2]
```diff
@@ -695,6 +695,15 @@ IProperties * CXmlEclElement::getProperties()
     //MORE: This should set individual properties rather than the "flags", or use flags defined in hqlexpr.hpp
     switch (type)
     {
+    case ESTdefinition:
```
Comment (ghalliday) [inline ecl/hql/hqlcollect.cpp]: Missing the default assignment of extraFlags

File: ecl/hql/hqlexpr.cpp [hunk 3]
```diff
@@ -914,14 +914,20 @@ void HqlParseContext::setGatherMeta(const MetaOptions & options)
...
     StringBuffer sillyTempBuffer;
     getFileContentText(sillyTempBuffer, contents);  // We can't rely on IFileContents->getText() being null terminated..
     target->setProp(prop, sillyTempBuffer);
 
     ISourcePath * sourcePath = contents->querySourcePath();
     target->setProp("@sourcePath", str(sourcePath));
+    if (checkDirty && contents->isDirty())
+    {
+        target->setPropBool("@dirty", true);
+        Owned<IError> error = createError(CategoryInformation, SeverityInformation, WRN_DEFINITION_SANDBOXED, "Definition has been modified", str(contents->querySourcePath()), 0, 0, 0);
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: not sure this is the correct place.  I will need to think about it.

File: ecl/hql/hqlexpr.ipp [hunk 4]
```diff
@@ -497,6 +497,53 @@ public:
...
+                    WARNLOG("Failed to run git status for %s", sourcePath->queryStr());
+                }
+                else
+                {
+                    try
+                    {
+                        unsigned retcode = pipe->wait();
+                        StringBuffer buf;
+                        Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+                        const size32_t chunkSize = 128;
```
Comment (ghalliday) [inline ecl/hql/hqlexpr.ipp]: another example of that duplicated code.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a few comments.  I'm not sure about the warning - it needs more thought.

File: ecl/eclcc/eclcc.cpp [hunk 5]
```diff
@@ -1498,7 +1500,7 @@ void EclCC::processFile(EclCompileInstance & instance)
     Owned<ISourcePath> sourcePath = optNoSourcePath ? NULL : createSourcePath(curFilename);
     Owned<IFileContents> queryText = createFileContentsFromFile(curFilename, sourcePath, false, NULL);
     const char * queryTxt = queryText->getText();
-    if (optArchive || optGenerateDepend || optSaveQueryArchive)
+    if (optArchive || optGenerateDepend || optSaveQueryArchive || optCheckDirty)
```
Comment (richardkchapman) [inline ecl/eclcc/eclcc.cpp]: It does because it uses the archive creation code to do the checking of version.

File: ecl/eclcc/eclcc.cpp [hunk 6]
```diff
@@ -1697,7 +1699,48 @@ void EclCC::generateOutput(EclCompileInstance & instance)
...
+                        WARNLOG("Failed to run git describe");
+                    }
+                    else
+                    {
+                        try
+                        {
+                            unsigned retcode = pipe->wait();
+                            StringBuffer buf;
+                            Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+                            const size32_t chunkSize = 128;
```
Comment (richardkchapman) [inline ecl/eclcc/eclcc.cpp]: Yes, I know - I did look to see if there was somewhere that matched I could call but existing ones were slightly different.

File: ecl/eclcc/eclcc.cpp (hunk 1 above)
Comment (richardkchapman) [inline ecl/eclcc/eclcc.cpp]: It does because it uses the archive creation code to do the checking of status.

File: ecl/hql/hqlexpr.cpp (hunk 3 above)
Comment (richardkchapman) [inline ecl/hql/hqlexpr.cpp]: I tried other places - but they ended up calling git status far too often. Putting it here I know that only files that end up in the archive get checked

File: ecl/eclcc/eclcc.hpp [hunk 7]
```diff
@@ -77,6 +77,7 @@ const char * const helpText[] = {
     "!   -b            Batch mode.  Each source file is processed in turn.  Output",
     "!                 name depends on the input filename",
     "!   -checkVersion Enable/disable ecl version checking from archives",
+    "    -checkDirty   Report any modified attributes using git status",
```
Comment (richardkchapman) [inline ecl/eclcc/eclcc.hpp]: I don't have strong feelings either way...

**Comment by richardkchapman** [comment]
@ghalliday Back to you

File: ecl/eclcc/eclcc.cpp (hunk 6 above)
Comment (richardkchapman) [inline ecl/eclcc/eclcc.cpp]: I raised a Jira to look into that later

File: ecl/eclcc/eclcc.hpp [hunk 8]
```diff
@@ -40,7 +40,7 @@ const char * const helpText[] = {
     "    -foption[=value] Set an ecl option (#option)",
     "    -main <ref>   Compile definition <ref> from the source collection",
     "    -syntax       Perform a syntax check of the ECL",
-    "    -platform=hthor Generate code for hthor executable",
+    "    -E            Output preprocessed ECL in xml archive form",
```
Comment (richardkchapman) [inline ecl/eclcc/eclcc.hpp]: Fixed

File: ecl/hql/hqlcollect.cpp (hunk 2 above)
Comment (richardkchapman) [inline ecl/hql/hqlcollect.cpp]: Fixed

File: ecl/hql/hqlexpr.cpp (hunk 3 above)
Comment (ghalliday) [inline ecl/hql/hqlexpr.cpp]: See suggestion in last commit in my gitdescribe branch

**Comment by ghalliday** [comment]
@richardkchapman see my gitdescribe branch for suggestion for code location

**Comment by richardkchapman** [comment]
@ghalliday Repushed to incorporate your change and remove unnecessary archive creation

**Pull request merged**